File: Aug14/Carl/Q9.py
import sys
class Solution(object):
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        total_sum = 0
        minSubArrayLen = sys.maxsize
        start = 0 
        end = 0 
        while end < len(nums):
            total_sum += nums[end]
            end += 1
            while total_sum >= s:
                minSubArrayLen = min(minSubArrayLen, end-start)
                total_sum -= nums[start]
                start += 1
        if minSubArrayLen == sys.maxsize:
            return 0
        return minSubArrayLen

S = Solution()
print(S.minSubArrayLen(10, [1,3,6,2,4,10]))

File: 2018/Aug14/Troy/Q9.java
class Solution {
    // O(n) solution
    public int minSubArrayLen(int s, int[] nums) {
        int len = nums.length;
        int temp_count = 0;
        int result = Integer.MAX_VALUE;
        int start = 0;
        for (int end = 0; end < len; end++) {
            temp_count += nums[end];
            if (temp_count >= s) {
                if (result > (end - start + 1)) {
                    result = end - start + 1;
                }
                temp_count -= nums[start];
                temp_count -= nums[end];
                start++;
                end--;
            }
        }

        return result == Integer.MAX_VALUE ? 0 : result;
    }
}
